util/topgen.py

```diff
@@ -584,12 +584,16 @@ def generate_flash(topcfg, out_path):
     gen_rtl.gen_rtl(IpBlock.from_path(str(hjson_path), []), str(rtl_path))
 
 
-def generate_top_only(top_only_list, out_path, topname):
+def generate_top_only(top_only_dict, out_path, topname, alt_hjson_path=""):
     log.info("Generating top only modules")
 
-    for ip in top_only_list:
-        hjson_path = Path(__file__).resolve(
-        ).parent / "../hw/top_{}/ip/{}/data/{}.hjson".format(topname, ip, ip)
+    for ip, reggen_only in top_only_dict.items():
+
+        if reggen_only and alt_hjson_path:
+            hjson_path = Path(alt_hjson_path) / f"{ip}.hjson"
```

Comment:
can someone more fluent in python let me know if there's a reason to use the "{}".format() format vs the f"{}" option?
it feels like the latter is more condensed, and in the event of having a lot of things to format is also easier to read.

am i missing something?

Reply:
If it's my code, it's because I didn't know about `f'...'` until Philipp told me about it :-)

Reply:
As a minor nit, I'd consider refactoring this so that the `if / else` branch picked the path to the directory and then some unconditional code afterwards added the `f'{ip}.hjson'`. It makes it a bit more obvious that we're always picking the same filename, but possibly looking in a different directory.

Also, it might help with the long lines :-)

(Same comment applies in the equivalent code in `_process_top`, although maybe it's easier to read there, so doesn't matter so much)

Reply:
updated! this does seem much easier to read.